app/services/onestop_id.rb

```diff
@@ -1,123 +1,153 @@
+        @name = string.split(COMPONENT_SEPARATOR)[2]
+      elsif geohash && name
+        # Filter geohash and name; validate later
+        @geohash = geohash_filter(geohash)
+        @name = name_filter(name)
+      else
+        raise ArgumentError.new('either a string or entity/geohash/name must be specified')
+      end
+      # Check valid OnestopID
+      is_a_valid_onestop_id, errors = self.validate(self.to_s)
+      if !is_a_valid_onestop_id
+        raise ArgumentError.new(errors.join(', '))
+      end
+      self
     end
-    self
-  end
 
-  def to_s
-    [@entity_prefix, @geohash, @name].join(COMPONENT_SEPARATOR)
-  end
+    def to_s
+      [self.class::PREFIX, @geohash, @name].join(COMPONENT_SEPARATOR)
+    end
 
-  def self.create_identifier(feed_onestop_id, entity_prefix, entity_id)
-    IDENTIFIER_TEMPLATE.expand(
-      feed_onestop_id: feed_onestop_id,
-      entity_prefix: entity_prefix,
-      entity_id: entity_id
-    ).to_s
-  end
+    def validate(value)
+      split = value.split(COMPONENT_SEPARATOR)
+      errors = []
+      errors << 'must not be empty' if value.blank?
+      errors << 'incorrect length' unless split.size == self.class::NUM_COMPONENTS
+      errors << 'invalid geohash' unless validate_geohash(split[1])
+      errors << 'invalid name' unless validate_name(split[2])
+      # errors << 'invalid suffix' unless validate_suffix(split[3])
+      return (errors.size == 0), errors
+    end
 
-  def self.validate_onestop_id_string(onestop_id, expected_entity_type: nil)
-    errors = []
-    is_a_valid_onestop_id = true
+    def validate_geohash(value)
+      !(value =~ GEOHASH_FILTER)
+    end
 
-    if onestop_id.blank?
-      return false, ['must not be blank']
+    def validate_name(value)
+      !(value =~ NAME_FILTER)
     end
 
-    if onestop_id.split(COMPONENT_SEPARATOR).length != 3
-      errors << 'must include 3 components separated by hyphens ("-")'
-      is_a_valid_onestop_id = false
+    def validate_prefix(value)
+      value == self.PREFIX
     end
 
-    if expected_entity_type && onestop_id.split(COMPONENT_SEPARATOR)[0] != ENTITY_TO_PREFIX[expected_entity_type]
-      errors << "must start with \"#{ENTITY_TO_PREFIX[expected_entity_type]}\" as its 1st component"
-      is_a_valid_onestop_id = false
-    elsif expected_entity_type == nil && !valid_component?(:entity_prefix, onestop_id.split(COMPONENT_SEPARATOR)[0])
-      errors << "must start with \"#{ENTITY_TO_PREFIX.values.join(' or ')}\" as its 1st component"
-      is_a_valid_onestop_id = false
+    def validate_suffix(value)
+      true
     end
 
-    if onestop_id.split(COMPONENT_SEPARATOR)[1].length == 0 || !valid_component?(:geohash, onestop_id.split(COMPONENT_SEPARATOR)[1])
-      errors << 'must include a valid geohash as its 2nd component'
-      is_a_valid_onestop_id = false
+    private
+
+    def name_filter(value)
+      value.downcase.gsub(NAME_TILDE, '~').gsub(NAME_FILTER, '')
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.